### PR TITLE
[feature] add support for HTTP PATCH

### LINKED ISSUE #3698: [feature] add PATCH support to RESTservlet
**Is your feature request related to a problem? Please describe.**

APIs make use of a number of different HTTP verbs `OPTION`, `HEAD`, `PUT`, `POST` and `GET` are already supported.

`PATCH` is not supported yet.

**Describe the solution you'd like**

```sh
curl -XPATCH http://localhost:8080/exist/apps/myapp/resource/x -H "Content-Type: application/json" -d '{"property1":"value1"}'
```

would work as expected.

For that existdb's RESTServlet must override the `service` method of `javax.servlet.http.HTTPServlet` in order to recognize the PATCH verb as well.

**Describe alternatives you've considered**

The `PUT` verb can be used as a fallback until this feature is added to Existdb's XqueryServlet

**Additional context**

Reference: https://cxf.apache.org/javadoc/latest/org/apache/cxf/transport/servlet/AbstractHTTPServlet.html#service-javax.servlet.ServletRequest-javax.servlet.ServletResponse-

### PR DESCRIPTION
**Pull request by line-o**
### Description:

adds support for HTTP verb PATCH

### Reference:

closes #3698

### Type of tests:

jUnit tests in PatchTest class

### CODE REVIEW

**Comment by line-o** [comment]
Codacy reports bogus problems it seems.
The imports it is whining about are necessary and not unused.

**Comment by line-o** [comment]
Well, turns out it is possible to make codacy happy.

**Comment by line-o** [comment]
Can someone restart the AppVeyor build, please? @dizzzz :pray:

File: exist-core/src/main/java/org/exist/http/RESTServer.java [hunk 1]
```diff
@@ -1162,61 +1191,66 @@ public void doDelete(final DBBroker broker, final Txn transaction, final String
...
-                }
-                return true;
+            servletPath = servletPath.removeLastSegment();
+            if (servletPath == XmldbURI.EMPTY_URI) {
+                break;
             }
         }
-        return false;
+
+        // no xquery binary file found
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/http/RESTServer.java:1236]: This should be moved up I think to line 1217.

File: exist-core/src/main/java/org/exist/http/servlets/EXistServlet.java [hunk 2]
```diff
@@ -505,4 +577,14 @@ protected void doPost(final HttpServletRequest req, final HttpServletResponse re
             }
         }
     }
+
+    @Override
+    protected void service(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/http/servlets/EXistServlet.java:582]: `final` arguments please

File: exist-core/src/test/java/org/exist/xquery/functions/request/PatchTest.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+import org.xmldb.api.DatabaseManager;
+import org.xmldb.api.base.Collection;
+import org.xmldb.api.base.XMLDBException;
+import org.xmldb.api.modules.BinaryResource;
+import org.xmldb.api.modules.XMLResource;
+
+import static org.apache.commons.codec.binary.Base64.encodeBase64String;
+
+/**
+ * test HTTP PATCH capabilities of ExistServlet
```
Comment (adamretter) [inline exist-core/src/test/java/org/exist/xquery/functions/request/PatchTest.java:52]: `test` -> `Test`

File: exist-core/src/test/java/org/exist/xquery/functions/request/PatchTest.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            final String actualResponse = new String(os.toByteArray());
+            assertThat(actualResponse, valueMatcher);
+        }
+    }
+
+    private void assertHTTPStatusCode (final int code, final HttpResponse response) {
+        assertEquals(code, response.getStatusLine().getStatusCode());
+    }
+
+    private void assertMethodNotAllowed (Request req) throws IOException {
```
Comment (adamretter) [inline exist-core/src/test/java/org/exist/xquery/functions/request/PatchTest.java:160]: final args please

File: exist-core/src/test/java/org/exist/xquery/functions/request/PatchTest.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        final String testData = "<a><b><c>hello</c></b></a>";
+
+        Request patch = Request.Patch(getCollectionRootUri() + "/" + XML_FILENAME)
+                .bodyByteArray(testData.getBytes(UTF_8), ContentType.TEXT_XML);
+
+        assertMethodNotAllowed(patch);
+    }
+
+    private void assertResponse(Request method, String expectedData) throws IOException {
+        final HttpResponse response = method.execute().returnResponse();
```
Comment (adamretter) [inline exist-core/src/test/java/org/exist/xquery/functions/request/PatchTest.java:142]: Do these HttpClients need to be closed after use?

File: exist-core/src/test/java/org/exist/xquery/functions/request/PatchTest.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+                .bodyByteArray(testData.getBytes(UTF_8), ContentType.TEXT_XML);
+
+        assertResponse(patch, testData);
+    }
+
+    @Test
+    public void patchString() throws IOException {
+        final String testData = "12345";
+
+        Request patch = Request.Patch(getCollectionRootUri() + "/" + XQUERY_FILENAME)
```
Comment (adamretter) [inline exist-core/src/test/java/org/exist/xquery/functions/request/PatchTest.java:115]: final variables please.

**Review by adamretter — COMMENTED** [review]
In general LGTM - just a few small improvements to be made please.

File: exist-core/src/main/java/org/exist/http/RESTServer.java (hunk 1 above)
Comment (line-o) [reply exist-core/src/main/java/org/exist/http/RESTServer.java:1236]: This is a safeguard against the while loop breaking while `resource` still being null.

File: exist-core/src/test/java/org/exist/xquery/functions/request/PatchTest.java (hunk 5 above)
Comment (line-o) [reply exist-core/src/test/java/org/exist/xquery/functions/request/PatchTest.java:142]: Might be the case, IDK. The test class that served as a template did not close them => `GetDataTest`.

File: exist-core/src/main/java/org/exist/http/servlets/EXistServlet.java (hunk 2 above)
Comment (line-o) [reply exist-core/src/main/java/org/exist/http/servlets/EXistServlet.java:582]: will add `final` were applicable

File: exist-core/src/main/java/org/exist/http/RESTServer.java [hunk 7]
```diff
@@ -1162,61 +1191,66 @@ public void doDelete(final DBBroker broker, final Txn transaction, final String
 
     private boolean checkForXQueryTarget(final DBBroker broker, final Txn transaction,
         final XmldbURI path, final HttpServletRequest request,
-        final HttpServletResponse response) throws PermissionDeniedException,
-        NotFoundException, IOException, BadRequestException {
-        
+        final HttpServletResponse response) throws PermissionDeniedException, IOException, BadRequestException {
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/http/RESTServer.java:1194]: NotFoundException is gone?

File: exist-core/src/main/java/org/exist/http/RESTServer.java [hunk 8]
```diff
@@ -1162,61 +1191,66 @@ public void doDelete(final DBBroker broker, final Txn transaction, final String
...
         final Collection collection = broker.getCollection(path);
-        if (collection == null) {
-            XmldbURI servletPath = path;
-            LockedDocument lockedDocument = null;
-            DocumentImpl resource = null;
-            // work up the url path to find an
-            // xquery resource
-            while (null == resource) {
-                // traverse up the path looking for xquery objects
+        // a collection is not executable
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/http/RESTServer.java:1202]: executable means "has executable flag' ? = user has access to Collection ?

File: exist-core/src/main/java/org/exist/http/RESTServer.java (hunk 7 above)
Comment (line-o) [reply exist-core/src/main/java/org/exist/http/RESTServer.java:1194]: yep

File: exist-core/src/main/java/org/exist/http/RESTServer.java (hunk 8 above)
Comment (line-o) [reply exist-core/src/main/java/org/exist/http/RESTServer.java:1202]: No, here executable means "is an xquery".

File: exist-core/src/main/java/org/exist/http/RESTServer.java (hunk 7 above)
Comment (line-o) [reply exist-core/src/main/java/org/exist/http/RESTServer.java:1194]: It made no sense so I was wondering why it was there and then Idea also told me it was not needed.

**Comment by line-o** [comment]
Ready to be merged.

File: exist-core/src/main/java/org/exist/http/RESTServer.java (hunk 8 above)
Comment (dizzzz) [reply exist-core/src/main/java/org/exist/http/RESTServer.java:1202]: confusing still...

File: exist-core/src/main/java/org/exist/http/servlets/EXistServlet.java [hunk 9]
```diff
@@ -177,6 +178,77 @@ protected void doPut(final HttpServletRequest request, final HttpServletResponse
...
 
+    protected void doPatch(final HttpServletRequest request, final HttpServletResponse response)
+            throws ServletException, IOException {
+        // first, adjust the path
+        String path = adjustPath(request);
+
+        // second, perform descriptor actions
+        final Descriptor descriptor = Descriptor.getDescriptorSingleton();
+        if (descriptor != null) {
+            // TODO: figure out a way to log PATCH requests with
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/http/servlets/EXistServlet.java:189]: TODO?

Comment (line-o) [reply exist-core/src/main/java/org/exist/http/servlets/EXistServlet.java:189]: Yeah... a _really_ old one

Comment (line-o) [reply exist-core/src/main/java/org/exist/http/servlets/EXistServlet.java:189]: Well, actually I added it because it was there for PUT too.

**Review by dizzzz — APPROVED** [review]

File: exist-core/src/main/java/org/exist/http/servlets/EXistServlet.java (hunk 9 above)
Comment (line-o) [reply exist-core/src/main/java/org/exist/http/servlets/EXistServlet.java:189]: But the replay log will be removed anyway, as we discussed yesterday

**Review by adamretter — APPROVED** [review]

**Pull request merged**